numpyro/svi.py

```diff
@@ -151,7 +177,17 @@ def elbo(param_map, model, guide, model_args, guide_args, kwargs, constrain_fn):
     """
     param_map = constrain_fn(param_map)
     guide_log_density, guide_trace = log_density(guide, guide_args, kwargs, param_map)
-    model_log_density, _ = log_density(replay(model, guide_trace), model_args, kwargs, param_map)
+    if is_autoguide:
```

Comment:
I am trying to understand why we need this. Could you elaborate on what goes wrong if we remove this `if` block?

Reply:
I have tried to elaborate the difference of autoguide/svi/hmc at [this comment](https://github.com/pyro-ppl/numpyro/pull/273#issuecomment-518078197). Here if we remove `if` block, final values of model will be filled by values from guide (which corresponds to base values of model), which is incorrect. We want to substitute these values from guide to base values of model. In particular (see [this test](https://github.com/pyro-ppl/numpyro/pull/273/files#diff-21afc849214aedbc4266eafd7e3f1483R117))
```
def model():
    x = sample('x', dist.TransformedDistribution(base_dist, transform))
```
will be interpreted as the following model in autoguide
```
def model():
    x_base = sample('x', base_dist)
    x = transform(x_base)
```

Reply:
That makes sense. I am concerned about leakage of abstraction here though. I think it is important that when we call `autoguide()` we generate samples that are within the model's support and not base values. It seems to me then that the right place to handle this is within the autoguide class themselves, but I will need to run this code myself to have a more concrete proposal. One way would be to use something like [CatTransform](https://github.com/pytorch/pytorch/blob/master/torch/distributions/transforms.py#L562) in `autoguide.__call__` to do this (my initial thought, not sure if it would handle all the cases).

It may also turn out that what you are doing in this PR is indeed the cleanest way of supporting autoguides, but I'll need to play around with the code to convince myself. Would it be possible to separate out the SVI related changes (your earlier PR) from the autoguide related changes and put up 2 PRs? We can get the former merged before discussing the latter set of changes. If its all tangled up, don't worry about it. I'll just need a bit longer to review this, and offer more concrete suggestions.

Reply:
Separating into 2 PRs is pretty straightforward so I'll do it to clarify things. The reason I made this PR is while refactoring init strategy, I realize that it is complicated to have dynamic support in autoguide. Change it leads to a change in SVI too. Maybe the test for SVI passed while test for autoguide failed made you think that I made a PR for SVI first, then added some commits to support autoguide later. It is indeed the other way around. :D

Reply:
Haha..thanks for clarifying!